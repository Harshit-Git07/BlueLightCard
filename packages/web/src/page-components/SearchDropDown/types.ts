export type SearchDropDownProps = {
  onSearchCompanyChange: (companyId: string, company: string) => void;
  onSearchCategoryChange: (categoryId: string, categoryName: string) => void;
};
