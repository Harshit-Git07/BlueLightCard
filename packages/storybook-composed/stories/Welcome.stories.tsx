import type { Meta, StoryObj } from '@storybook/react';
import WelcomeComponent from './Welcome';

const meta: Meta<typeof WelcomeComponent> = {
  component: WelcomeComponent,
};

export default meta;

type Story = StoryObj<typeof WelcomeComponent>;

export const Welcome: Story = {
  args: {},
  argTypes: {},
};
