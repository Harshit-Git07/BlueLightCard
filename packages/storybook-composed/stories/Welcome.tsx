import React, { FC } from 'react';

const WelcomeComponent: FC = () => (
  <div>
    <p>
      Welcome to the Storybook Composed instance. Here you can find all of our stories across Web,
      Mobile Hybrid, and Shared UI for every brand. Please view these stories from the sidebar.
    </p>
  </div>
);

export default WelcomeComponent;
