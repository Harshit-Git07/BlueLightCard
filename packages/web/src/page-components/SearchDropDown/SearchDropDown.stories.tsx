import { Meta, StoryFn } from '@storybook/react';
import { SearchDropDownPresenter } from './SearchDropDown';

const componentMeta: Meta<typeof SearchDropDownPresenter> = {
  title: 'Components/SearchDropDown',
  component: SearchDropDownPresenter,
  argTypes: {},
  parameters: {
    controls: {
      exclude: [
        'categories',
        'companies',
        'isLoading',
        'onSearchCategoryChange',
        'onSearchCompanyChange',
      ],
    },
  },
};

const mockData = {
  data: {
    response: {
      categories: [
        {
          id: '11',
          name: 'Children and toys',
          __typename: 'CategoryMenu',
        },
        {
          id: '3',
          name: 'Days out',
          __typename: 'CategoryMenu',
        },
        {
          id: '1',
          name: 'Electrical and phones',
          __typename: 'CategoryMenu',
        },
        {
          id: '6',
          name: 'Entertainment',
          __typename: 'CategoryMenu',
        },
        {
          id: '8',
          name: 'Fashion',
          __typename: 'CategoryMenu',
        },
        {
          id: '16',
          name: 'Featured',
          __typename: 'CategoryMenu',
        },
        {
          id: '9',
          name: 'Financial and insurance',
          __typename: 'CategoryMenu',
        },
        {
          id: '15',
          name: 'Food and drink',
          __typename: 'CategoryMenu',
        },
        {
          id: '5',
          name: 'Gifts',
          __typename: 'CategoryMenu',
        },
        {
          id: '4',
          name: 'Health and beauty',
          __typename: 'CategoryMenu',
        },
        {
          id: '2',
          name: 'Holidays and travel',
          __typename: 'CategoryMenu',
        },
        {
          id: '10',
          name: 'Home and garden',
          __typename: 'CategoryMenu',
        },
        {
          id: '7',
          name: 'Motoring',
          __typename: 'CategoryMenu',
        },
      ],
      companies: [
        {
          id: '26529',
          name: ' Youth & Earth',
          __typename: 'CompanyMenu',
        },
        {
          id: '14779',
          name: '1 Stop Barber Shop ',
          __typename: 'CompanyMenu',
        },
        {
          id: '15939',
          name: '11 Degrees',
          __typename: 'CompanyMenu',
        },
        {
          id: '23876',
          name: '18montrose',
          __typename: 'CompanyMenu',
        },
        {
          id: '20128',
          name: '1st 4 Signs Ltd',
          __typename: 'CompanyMenu',
        },
        {
          id: '13571',
          name: '360 Play',
          __typename: 'CompanyMenu',
        },
        {
          id: '5961',
          name: '361 Property Management ',
          __typename: 'CompanyMenu',
        },
        {
          id: '15943',
          name: '3RD ROCK',
          __typename: 'CompanyMenu',
        },
        {
          id: '19750',
          name: '43T Clothing',
          __typename: 'CompanyMenu',
        },
      ],
    },
  },
};

const DefaultTemplate: StoryFn<typeof SearchDropDownPresenter> = (args) => (
  <SearchDropDownPresenter {...args}></SearchDropDownPresenter>
);

export const Default = DefaultTemplate.bind({});

Default.args = {
  categories: mockData.data.response.categories,
  companies: mockData.data.response.companies,
  isLoading: false,
  onSearchCategoryChange: () => {},
  onSearchCompanyChange: () => {},
};

export default componentMeta;
