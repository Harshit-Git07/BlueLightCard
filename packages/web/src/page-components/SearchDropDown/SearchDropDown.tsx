import React, { FC, useContext } from 'react';
import { SearchDropDownProps } from './types';
import Link from '@/components/Link/Link';
import InputTypeSelect from '@/components/InputTypeSelect/InputTypeSelect';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faLocationDot } from '@fortawesome/pro-solid-svg-icons';
import { PillGroup } from '@bluelightcard/shared-ui';
import UserContext from '@/context/User/UserContext';
import useFetchCompaniesOrCategories, {
  CategoryType,
  CompanyType,
} from '@/hooks/useFetchCompaniesOrCategories';

type SearchDropDownPresenterProps = {
  onSearchCategoryChange: (categoryId: string, categoryName: string) => void;
  onSearchCompanyChange: (companyId: string, companyName: string) => void;
  categories: CategoryType[];
  companies: CompanyType[];
  isLoading: boolean;
};

export const SearchDropDownPresenter = ({
  categories,
  companies,
  isLoading,
  onSearchCategoryChange,
  onSearchCompanyChange,
}: SearchDropDownPresenterProps) => {
  const changeCategoryHandler = (categoryId: string, categoryName: string) => {
    onSearchCategoryChange(categoryId, categoryName);
  };

  const changeCompanyHandler = (companyId: string, companyName: string) => {
    onSearchCompanyChange(companyId, companyName);
  };

  return (
    <div className="mx-auto pt-[54px] pb-[100px] max-w-[1408px]">
      <div className="flex gap-2 px-4 py-[10px]">
        <Link href={`/nearme.php`} useLegacyRouting>
          <FontAwesomeIcon icon={faLocationDot} size="lg" className="mr-2" />
          <span className="font-typography-body font-typography-body-weight text-typography-body leading-typography-body tracking-typography-body">
            Offers near you
          </span>
        </Link>
      </div>
      <div className="my-[22px] px-4">
        <PillGroup
          title={'Browse categories'}
          pillGroup={categories.map((cat) => ({
            id: Number(cat.id),
            label: cat.name,
          }))}
          onSelectedPill={(id) => {
            const selectedCategory = categories.find((cat) => cat.id === id.toString());
            if (selectedCategory) {
              changeCategoryHandler(selectedCategory.id, selectedCategory.name);
            }
          }}
        />
      </div>
      <div className="px-4">
        <p className="mb-2 font-typography-title-large font-typography-title-large-weight text-typography-title-large leading-typography-title-large tracking-typography-title-large text-colour-onSurface-light dark:text-colour-onSurface-dark">
          Browse companies
        </p>
        <InputTypeSelect
          options={companies.map((cat) => ({
            id: cat.id,
            label: cat.name,
          }))}
          placeholder={'Select company'}
          onSelect={(id: string) => {
            const selectedCategory = companies.find((comp) => comp.id === id);
            if (selectedCategory) {
              changeCompanyHandler(selectedCategory.id, selectedCategory.name);
            }
          }}
        />
      </div>
    </div>
  );
};

const SearchDropDown: FC<SearchDropDownProps> = ({
  onSearchCategoryChange,
  onSearchCompanyChange,
}) => {
  const userCtx = useContext(UserContext);
  const { categories, companies, isLoading } = useFetchCompaniesOrCategories(userCtx);

  return (
    <SearchDropDownPresenter
      isLoading={isLoading}
      companies={companies}
      categories={categories}
      onSearchCategoryChange={onSearchCategoryChange}
      onSearchCompanyChange={onSearchCompanyChange}
    />
  );
};

export default SearchDropDown;
