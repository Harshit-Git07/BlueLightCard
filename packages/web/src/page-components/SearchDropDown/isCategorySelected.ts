export const isCategorySelected = (categoryId: string, pathname: string): boolean => {
  if (!pathname.startsWith('/offers.php?cat=true&type=')) {
    return false;
  }
  return pathname.includes(categoryId);
};
