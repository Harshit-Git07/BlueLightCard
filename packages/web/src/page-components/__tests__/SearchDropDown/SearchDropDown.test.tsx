import { act, render, fireEvent, within, cleanup } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { composeStories } from '@storybook/react';
import * as stories from '../../SearchDropDown/SearchDropDown.stories';

const { Default, Interaction } = composeStories(stories);

describe('SearchDropDown', () => {
  beforeEach(() => {
    window.HTMLElement.prototype.scrollIntoView = jest.fn();
  });

  it('renders the dropdown', () => {
    const { container } = render(<Default />);

    expect(container).toMatchSnapshot();
  });

  it('is interactive', async () => {
    const onSearchCategoryChange = jest.fn();
    const onSearchCompanyChange = jest.fn();
    const onClose = jest.fn();

    const { container } = render(
      <Interaction
        onSearchCategoryChange={onSearchCategoryChange}
        onSearchCompanyChange={onSearchCompanyChange}
        onClose={onClose}
      />
    );

    const category = within(container).getByText('Children and toys');
    fireEvent.click(category);

    const companySearch = within(container).getByPlaceholderText('Search for a company');
    await act(() => userEvent.type(companySearch, 'you'));

    const company = await within(container).findByText('Youth & Earth');
    fireEvent.click(company);

    const overlay = within(container).getByTestId('search-dropdown-overlay');
    fireEvent.click(overlay);

    expect(onSearchCategoryChange).toHaveBeenCalledWith('11', 'Children and toys');
    expect(onSearchCompanyChange).toHaveBeenCalledWith('26529', 'Youth & Earth');
    expect(onClose).toHaveBeenCalled();
  });

  it('closes the SearchDropDown when clicking on a Category', async () => {
    const onSearchCategoryChange = jest.fn();
    const mockOnClose = jest.fn();

    const { container } = render(
      <Interaction
        isOpen={true}
        onSearchCategoryChange={onSearchCategoryChange}
        onSearchCompanyChange={() => {}}
        onClose={mockOnClose}
      />
    );

    const category = within(container).getByText('Children and toys');
    fireEvent.click(category);

    const overlay = within(container).getByTestId('search-dropdown-overlay');
    fireEvent.click(overlay);

    expect(onSearchCategoryChange).toHaveBeenCalledWith('11', 'Children and toys');
    expect(mockOnClose).toHaveBeenCalled();
  });

  it('closes the SearchDropDown when clicking on a Company', async () => {
    const onSearchCompanyChange = jest.fn();
    const mockOnClose = jest.fn();

    const { container } = render(
      <Interaction
        onSearchCategoryChange={() => {}}
        onSearchCompanyChange={onSearchCompanyChange}
        onClose={mockOnClose}
      />
    );

    const companySearch = within(container).getByPlaceholderText('Search for a company');
    await act(() => userEvent.type(companySearch, 'you'));

    const company = await within(container).findByText('Youth & Earth');
    fireEvent.click(company);

    const overlay = within(container).getByTestId('search-dropdown-overlay');
    fireEvent.click(overlay);

    expect(onSearchCompanyChange).toHaveBeenCalledWith('26529', 'Youth & Earth');
    expect(mockOnClose).toHaveBeenCalled();
  });
});

describe('SearchDropDown scroll behavior', () => {
  const onSearchCategoryChange = jest.fn();
  const onSearchCompanyChange = jest.fn();
  const onClose = jest.fn();

  afterEach(() => {
    cleanup();
    document.body.classList.remove('no-scroll');
  });

  it('adds the no-scroll class to the body when isOpen is true', () => {
    render(
      <Default
        isOpen={true}
        onSearchCategoryChange={onSearchCategoryChange}
        onSearchCompanyChange={onSearchCompanyChange}
        onClose={onClose}
      />
    );

    expect(document.body.classList.contains('no-scroll')).toBe(true);
  });

  it('removes the no-scroll class from the body when isOpen is false', () => {
    render(
      <Default
        isOpen={false}
        onSearchCategoryChange={onSearchCategoryChange}
        onSearchCompanyChange={onSearchCompanyChange}
        onClose={onClose}
      />
    );

    expect(document.body.classList.contains('no-scroll')).toBe(false);
  });

  it('removes the no-scroll class from the body when component unmounts', () => {
    const { unmount } = render(
      <Default
        isOpen={true}
        onSearchCategoryChange={onSearchCategoryChange}
        onSearchCompanyChange={onSearchCompanyChange}
        onClose={onClose}
      />
    );

    expect(document.body.classList.contains('no-scroll')).toBe(true);

    unmount();

    expect(document.body.classList.contains('no-scroll')).toBe(false);
  });
});
