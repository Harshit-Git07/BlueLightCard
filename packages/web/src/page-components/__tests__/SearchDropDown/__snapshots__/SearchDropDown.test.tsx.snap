// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchDropDown should match snapshot 1`] = `
<div
  className="mx-auto pt-[54px] pb-[100px] max-w-[1408px]"
>
  <div
    className="flex gap-2 px-4 py-[10px]"
  >
    <a
      className="text-colour-primary dark:text-colour-primary-dark"
      data-testid="anchor-link"
      href="/nearme.php"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-location-dot fa-lg mr-2"
        data-icon="location-dot"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={{}}
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M215.7 499.2C267 435 384 279.4 384 192C384 86 298 0 192 0S0 86 0 192c0 87.4 117 243 168.3 307.2c12.3 15.3 35.1 15.3 47.4 0zM192 128a64 64 0 1 1 0 128 64 64 0 1 1 0-128z"
          fill="currentColor"
          style={{}}
        />
      </svg>
      <span
        className="font-typography-body font-typography-body-weight text-typography-body leading-typography-body tracking-typography-body"
      >
        Offers near you
      </span>
    </a>
  </div>
  <div
    className="my-[22px] px-4"
  >
    <div
      className="w-full"
    >
      <p
        className="py-1.5 mb-2 font-typography-title-large font-typography-title-large-weight text-typography-title-large leading-typography-title-large tracking-typography-title-large text-colour-onSurface-light dark:text-colour-onSurface-dark"
      >
        Browse categories
      </p>
      <div
        className="flex gap-[15px] flex-wrap"
      />
    </div>
  </div>
  <div
    className="px-4"
  >
    <p
      className="mb-2 font-typography-title-large font-typography-title-large-weight text-typography-title-large leading-typography-title-large tracking-typography-title-large text-colour-onSurface-light dark:text-colour-onSurface-dark"
    >
      Browse companies
    </p>
    <div
      className="relative inline-block w-full"
    >
      <div
        className="rounded-[4px] bg-colour-surface-light dark:bg-colour-surface-dark border-colour-onSurface-outline-light dark:border-colour-onSurface-outline-dark border px-4 py-3 cursor-pointer flex justify-between items-center"
        onClick={[Function]}
      >
        <span
          className="font-typography-body text-typography-body font-typography-body-weight leading-typography-body tracking-typography-body text-colour-onSurface-subtle-light dark:text-colour-onSurface-subtle-dark"
        >
          Select company
        </span>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-chevron-down fa-sm text-colour-onSurface-subtle-light dark:text-colour-onSurface-subtle-dark font-typography-body text-typography-body font-typography-body-weight leading-typography-body tracking-typography-body"
          data-icon="chevron-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={{}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
            fill="currentColor"
            style={{}}
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
